packages/@lwc/engine-core/src/framework/context.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isUndefined, ArrayPush } from '@lwc/shared';
+import { guid } from './utils';
+import { VM, getAssociatedVMIfPresent } from './vm';
+import {
+    WireAdapterConstructor,
+    ContextValue,
+    getAdapterToken,
+    setAdapterToken,
+    WireDef,
+} from './wiring';
+
+export type WireContextSubscriptionCallback = (
+    subscriptionPayload: WireContextSubscriptionPayload
+) => void;
+
+export interface WireContextSubscriptionPayload {
+    setNewContext(newContext: ContextValue): void;
+    setDisconnectedCallback(disconnectCallback: () => void): void;
+}
+
+interface ContextConsumer {
+    provide(newContext: ContextValue): void;
+}
+
+interface ContextProviderOptions {
+    consumerConnectedCallback: (consumer: ContextConsumer) => void;
+    consumerDisconnectedCallback?: (consumer: ContextConsumer) => void;
+}
+
+export function createContextProvider(adapter: WireAdapterConstructor) {
+    let adapterContextToken = getAdapterToken(adapter);
+    if (!isUndefined(adapterContextToken)) {
+        throw new Error(`Adapter already has a context provider.`);
+    }
+    adapterContextToken = guid();
+    setAdapterToken(adapter, adapterContextToken);
+    const providers = new WeakSet<EventTarget>();
+
+    return (elmOrComponent: EventTarget, options: ContextProviderOptions) => {
+        if (providers.has(elmOrComponent)) {
+            throw new Error(`Adapter was already installed on ${elmOrComponent}.`);
+        }
+
+        const vm = getAssociatedVMIfPresent(elmOrComponent);
+        if (!vm) {
```

Comment:
```suggestion
        if (isUndefined(vm)) {
```

I know we're inconsistent about this, and I'm not even sure why we do it (micro-optimization?), but we should try to be consistent about it probably.